wire-library/wire-grpc-client/src/jvmMain/kotlin/com/squareup/wire/internal/grpc.kt

```diff
@@ -115,25 +129,40 @@ internal fun <R : Any> SendChannel<R>.readFromResponseBodyCallback(
  * 1. read a message (non blocking, suspending code)
  * 2. write it to the stream (blocking)
  * 3. repeat. We also have to wait for all 2s to end before closing the writer
+ *
+ * If this fails reading a message from the channel, we need to call [MessageSink.cancel]. If it
+ * fails writing a message to the network, we shouldn't call that method.
+ *
+ * If it fails either reading or writing we need to call [MessageSink.close] (via [Closeable.use])
+ * and [ReceiveChannel.cancel].
  */
 internal suspend fun <S : Any> ReceiveChannel<S>.writeToRequestBody(
   requestBody: PipeDuplexRequestBody,
   minMessageToCompress: Long,
   requestAdapter: ProtoAdapter<S>,
   callForCancel: Call
 ) {
-  requestBody.messageSink(minMessageToCompress, requestAdapter, callForCancel)
-    .use { requestWriter ->
-      var success = false
+  val requestWriter = requestBody.messageSink(minMessageToCompress, requestAdapter, callForCancel)
+  try {
+    requestWriter.use {
+      var channelReadFailed = true
       try {
         consumeEach { message ->
+          channelReadFailed = false
           requestWriter.write(message)
+          channelReadFailed = true
```

Comment:
I don't get this line. `consumeEach` is throwing after the _whole_ block has been executed? So if it doesn't throw, the line below override `channelReadFailed` to `false` so we're good?

Reply:
If we were to inline and desugar consumeEach(), it’d looks something like this:

```
  while (true) {
    val message = receiveOrNull()
    if (message == null) break
   requestWriter.write(message)
  }
```

Given that behavior, I want to detect exceptions only thrown by receiveOrNull():

```
  var channelReadFailed: Boolean
  while (true) {
    channelReadFailed = true
    val message = receiveOrNull()
    channelReadFailed = false
    if (message == null) break
    requestWriter.write(message)
  }
```

But unfortunately due to the way consumeEach works, I can’t slot my code in there. So instead I rewrite the above to this:

```
  var channelReadFailed: Boolean = true
  while (true) {
    val message = receiveOrNull()
    if (message == null) break
    channelReadFailed = false
    requestWriter.write(message)
    channelReadFailed = true
  }
  channelReadFailed = false
```

I hope you agree that this is equivalent?

Reply:
Perfect yes, thank you